boofuzz/tcp_socket_connection.py

```diff
@@ -0,0 +1,144 @@
+        recv_timeout (float): Seconds to wait for recv before timing out. Default 5.0.
+        server (bool): Set to True to enable server side fuzzing.
+
+    """
+
+    def __init__(self, host, port, send_timeout=5.0, recv_timeout=5.0, server=False):
+        super(TCPSocketConnection, self).__init__(send_timeout, recv_timeout)
+
+        self.host = host
+        self.port = port
+        self.server = server
+        self._serverSock = None
+
+    def close(self):
+        super(TCPSocketConnection, self).close()
+
+        if self.server:
+            self._serverSock.close()
+
+    def open(self):
+        self._sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+
+        # call superclass to set timeout sockopt
+        super(TCPSocketConnection, self).open()
+
+        if self.server:
+            self._sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+            try:
+                self._sock.bind((self.host, self.port))
+            except socket.error as e:
+                if e.errno == errno.EADDRINUSE:
+                    raise exception.BoofuzzOutOfAvailableSockets()
+                else:
+                    raise
+
+            self._serverSock = self._sock
+            try:
+                self._serverSock.listen(1)
+                self._sock, addr = self._serverSock.accept()
+            except socket.error as e:
+                # When connection timeout expires, tear down the server socket so we can re-open it again after
+                # restarting the target.
+                self.close()
+                if e.errno in [errno.EAGAIN]:
+                    raise exception.BoofuzzTargetConnectionFailedError(str(e))
+                else:
+                    raise
+
+        try:
+            self._sock.connect((self.host, self.port))
+        except socket.error as e:
+            if e.errno == errno.EADDRINUSE:
+                raise exception.BoofuzzOutOfAvailableSockets()
+            elif e.errno in [errno.ECONNREFUSED, errno.EINPROGRESS, errno.ETIMEDOUT]:
+                raise exception.BoofuzzTargetConnectionFailedError(str(e))
+            else:
+                raise
+
+    def recv(self, max_bytes):
+        """
+        Receive up to max_bytes data from the target.
+
+        Args:
+            max_bytes (int): Maximum number of bytes to receive.
+
+        Returns:
+            Received data.
+        """
+        data = b""
+
+        try:
+            data = self._sock.recv(max_bytes)
+        except socket.timeout:
+            data = b""
+        except socket.error as e:
+            # Issue 348, PR 361: SSLError handling on python < 3.3
+            # Before Python 3.3, SSLError used to be a subtype of socket.error.
+            # To ensure it is properly handled in SSLSocketConnection, we have to
+            # re-raise it here if we're handling a ssl.SSLErorr.
+            if type(e) is ssl.SSLError:
+                raise
```

Comment:
I don't think this is needed. We already cover this case by `else: raise`

Reply:
I‘m not too sure either; however I wanted to play it safe when merging. I couldn’t figure out a quick unit test for this case, so I decided to go the explicit route and raise as early as possible while documenting why.

If we have a testcase that demonstrates that this isn‘t a problem with the new structure, we should remove that bit of course.

Reply:
I did some manual testing and both `ssl.SSLError` and `socket.error` both inherit from `OSError`. According to [this](https://docs.python.org/3/library/socket.html#socket.error), `socket.error` is just a depreciated alias for `OSError`. So catching `socket.error` will always also catch `ssl.SSLError`.
I verified this by some manual testing on Python 2.7 and 3.7 with a debugger.

I also ran some tests with
```python
if type(e) is ssl.SSLError:
    raise
```
and found out that the type of my provoked ssl error was of the type `SSLCertVerificationError` which doesn't match `ssl.SSLError`.

In conclusion, it's safe to remove that code.